keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,859 @@
+
+
+### Story 2
+
+Prevent Pods from being scheduled on nodes lacking their required Kubelet, Kernel, or Runtime features, avoiding errors, degraded performance, or silent failures.
+
+**Examples:**
+
+*   Linux kernel can be compiled with support for both AppArmor and SELinux security modules, only one can be actively enforced at any given time. This is a key reason why Kubernetes clusters can exhibit heterogeneity in their security configuration, making it crucial for the `kube-scheduler` to know which Linux Security Module is active on each node (typically via Node Labels) for pod placement. Instead, kubelet can expose this kernel capability via NodeCapabilities and the kube-scheduler can infer this from the pod spec (`securitycontext.seLinuxOptions,securitycontext.appArmorProfile`) and match it to the right node.
+*  [PodLevelResources](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pod-level-resources/) is currently not supported on Windows nodes. The capability advertised by kubelet can be dependent on both the feature gate and the operating system which the kube-scheduler can filter out all Windows nodes.
+*   When pods request for a specific [RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) (`spec.runtimeClassName`), we currently use node selectors to make sure they are on appropriate nodes . Instead, kube-scheduler can automatically identify nodes that possess the corresponding capability.
+
+## Design Details
+
+### API Changes
+
+ \
+Add a field `Capabilities` field as type `map[string]string` to the `Node.Status` structure. 
+
+```
+type NodeStatus struct {
+    // ... existing fields
+    // Capabilities provides a structured way to report various capabilities of the node. Keys are typically DNS-style and values are strings.
+    // +optional 
+    // +featureGate=NodeCapabilities
+    Capabilities map[string]string `json:"capabilities,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+Capabilities:
+   feature.kubernetes.io/inPlacePodResize: "true"
+   feature.kubernetes.io/podLevelResourcesSupport: "true"
+   feature.kubernetes.io/swapSupport: "true"
+   runtime.kubernetes.io/handlers: "runc, runsc"
+   kernel.kubernetes.io/LSMConfig: "seLinux"   
```

Comment:
rolling out the LMSConfig as capability will be a big challenge. Even if we ignore the desire to declare this as a node-restricted label in many environments, backfilling this and eventually enforcing with many "incompatible" configuration that might be existing today, will be a challenge